### PR TITLE
[wasm] Fix trimming errors for library tests, to help building with AOT

### PR DESCRIPTION
**Pull request by radical**

### CODE REVIEW

File: eng/testing/ildescriptors/ILLink.Descriptor.Castle.xml [hunk 1]
```diff
@@ -0,0 +1,13 @@
+<linker>
```
Comment (eerhardt) [inline eng/testing/ildescriptors/ILLink.Descriptor.Castle.xml:1]: All these files should have `Descriptors` in the name following the naming pattern set by the linker.

See the docs here:
https://github.com/mono/linker/blob/55bd0ebd00ddbf2cd817314b1cf6824003cfb463/docs/data-formats.md#descriptor-format

File: eng/testing/ildescriptors/ILLink.Descriptor.Serialization.xml [hunk 2]
```diff
@@ -0,0 +1,10 @@
+<linker>
+  <assembly fullname="System.Private.DataContractSerialization">
+    <type fullname="System.Runtime.Serialization.KeyValue`2" />
```
Comment (eerhardt) [inline eng/testing/ildescriptors/ILLink.Descriptor.Serialization.xml:3]: cc @joperezr - is your change going to make it so this is no longer necessary?

Comment (eerhardt) [reply eng/testing/ildescriptors/ILLink.Descriptor.Serialization.xml:3]: In general, I don't think we should be adding `ILLink.Descriptors` files for our own assemblies. Instead we should be fixing our assemblies to be trimmable.

File: eng/testing/ildescriptors/ILLink.Descriptor.Serialization.xml [hunk 3]
```diff
@@ -0,0 +1,10 @@
...
+    <type fullname="System.Runtime.Serialization.DateTimeOffsetAdapter" />
+    <type fullname="System.Runtime.Serialization.KeyValuePairAdapter`2" />
+  </assembly>
+  <assembly fullname="System.Private.Xml.Linq">
+    <type fullname="System.Xml.Linq.XElement" />
```
Comment (eerhardt) [inline eng/testing/ildescriptors/ILLink.Descriptor.Serialization.xml:8]: Why is this necessary? It shouldn't be.

File: src/libraries/System.Runtime.Serialization.Xml/tests/ILLink.Descriptor.xml [hunk 4]
```diff
@@ -0,0 +1,7 @@
+<linker>
+  <assembly fullname="System.Data.Common">
+    <type fullname="System.Data.DataColumn">
+      <method signature="System.Void .ctor()" />
```
Comment (eerhardt) [inline src/libraries/System.Runtime.Serialization.Xml/tests/ILLink.Descriptor.xml:4]: cc @krwq - is your change going to fix this?

File: src/libraries/System.Reflection/tests/ILLink.Descriptor.xml [hunk 5]
```diff
@@ -0,0 +1,7 @@
+<linker>
+  <assembly fullname="System.Private.CoreLib">
+      <type fullname="System.Collections.ArrayList" />
```
Comment (eerhardt) [inline src/libraries/System.Reflection/tests/ILLink.Descriptor.xml:3]: Why is this necessary? It shouldn't be.

File: src/libraries/Microsoft.Extensions.Logging/tests/Common/ILLink.Descriptor.xml [hunk 6]
```diff
@@ -0,0 +1,7 @@
+<linker>
+  <assembly fullname="System.Diagnostics.TraceSource">
+    <type fullname="System.Diagnostics.TraceListener">
+        <method signature="System.Void .ctor()" />
```
Comment (eerhardt) [inline src/libraries/Microsoft.Extensions.Logging/tests/Common/ILLink.Descriptor.xml:4]: Do you know why this is necessary? It shouldn't be.

File: src/libraries/System.Reflection/tests/ILLink.Descriptor.xml (hunk 5 above)
Comment (radical) [reply src/libraries/System.Reflection/tests/ILLink.Descriptor.xml:3]: I was seeing this as:
```
02:19:20] fail: [FAIL] System.Reflection.Tests.TypeInfoTests.IsAssignable(type: typeof(System.Collections.IList), c: typeof(System.Collections.ArrayList), expected: True)
[02:19:20] info: Assert.Equal() Failure
[02:19:20] info: Expected: True
[02:19:20] info: Actual:   False
[02:19:20] info:    at System.Reflection.Tests.TypeInfoTests.IsAssignable(Type type, Type c, Boolean expected)
[02:19:20] info:    at System.Reflection.RuntimeMethodInfo.Invoke(Object obj, BindingFlags invokeAttr, Binder binder, Object[] parameters, CultureInfo culture)
```
https://github.com/radical/runtime/commit/3cdf08a0ece59578afabc3ffb661de8f7a07d122

This PR is an attempt to remove the trimming errors, to enable AOT tests to be built. I didn't delve into why the linker wasn't being able to pick up the usage. We could look into that, when I make a clean PR based on `main`.

File: src/libraries/Microsoft.Extensions.Logging/tests/Common/ILLink.Descriptors.xml [hunk 7]
```diff
@@ -0,0 +1,7 @@
+<linker>
+  <assembly fullname="System.Diagnostics.TraceSource">
```
Comment (marek-safar) [inline src/libraries/Microsoft.Extensions.Logging/tests/Common/ILLink.Descriptors.xml:2]: This one is a bit unexpected, do you know where is it used?

File: src/libraries/System.Reflection.MetadataLoadContext/tests/ILLink.Descriptors.xml [hunk 8]
```diff
@@ -0,0 +1,3 @@
+<linker>
+  <assembly fullname="mscorlib" />
```
Comment (eerhardt) [inline src/libraries/System.Reflection.MetadataLoadContext/tests/ILLink.Descriptors.xml:2]: Is this still necessary with the latest linker? We had a few fixes that went in to preserve facade assemblies.

File: src/libraries/System.Reflection/tests/CoreCLR/System.Reflection.CoreCLR.Tests.csproj [hunk 9]
```diff
@@ -14,5 +14,7 @@
   </ItemGroup>
   <ItemGroup>
     <ProjectReference Include="..\TestExe\System.Reflection.TestExe.csproj" />
+
+    <TrimmerRootDescriptor Include="$(MSBuildThisFileDirectory)..\ILLink.Descriptors.CoreCLR.xml" />
```
Comment (eerhardt) [inline src/libraries/System.Reflection/tests/CoreCLR/System.Reflection.CoreCLR.Tests.csproj:18]: Can this `ILLink.Descriptors.CoreCLR.xml` file move next to the .csproj that needs it?

File: eng/testing/tests.props [hunk 10]
```diff
@@ -6,6 +6,7 @@
     <TestDependsOn Condition="'$(TestNoBuild)' != 'true' and '$(BuildAllProjects)' != 'true'">Build</TestDependsOn>
     <TestDependsOn>$(TestDependsOn);GenerateRunScript;RunTests</TestDependsOn>
     <VSTestNoLogo>true</VSTestNoLogo>
+    <ILDescriptorsPath>$(MSBuildThisFileDirectory)ildescriptors\</ILDescriptorsPath>
```
Comment (eerhardt) [inline eng/testing/tests.props:9]: This needs a better name. How about?

`<ILLinkDescriptorsPath>$(MSBuildThisFileDirectory)ILLinkDescriptors\</ILLinkDescriptorsPath>`

**Comment by eerhardt** [comment]
> <linker>

This file should go in the `ILLinkDescriptors` folder.

---
Refers to: eng/testing/ILLink.Descriptors.TestUtilities.xml:1 in e8cb433. [](commit_id = e8cb4334bc2a5faa11496354dac017a5d04e5355, deletion_comment = False)

File: src/libraries/Microsoft.Extensions.Logging/tests/Common/ILLink.Descriptors.xml (hunk 7 above)
Comment (eerhardt) [reply src/libraries/Microsoft.Extensions.Logging/tests/Common/ILLink.Descriptors.xml:2]: Agreed - this shouldn't be necessary because all the ILLink warnings are fixed in this assembly. This might be a product bug somewhere.

File: eng/testing/ildescriptors/ILLink.Descriptors.Serialization.xml [hunk 11]
```diff
@@ -0,0 +1,10 @@
+<linker>
+  <assembly fullname="System.Private.DataContractSerialization">
+    <type fullname="System.Runtime.Serialization.KeyValue`2" />
```
Comment (eerhardt) [inline eng/testing/ildescriptors/ILLink.Descriptors.Serialization.xml:3]: Are these still necessary on the latest main? @joperezr addressed the ILLink warnings in this assembly, so these might be fixed.

If they aren't fixed, we should open a bug to make sure we don't have a product bug here.

File: src/libraries/System.Reflection/tests/ILLink.Descriptors.xml [hunk 12]
```diff
@@ -0,0 +1,7 @@
+<linker>
+  <assembly fullname="System.Private.CoreLib">
+      <type fullname="System.Collections.ArrayList" />
```
Comment (eerhardt) [inline src/libraries/System.Reflection/tests/ILLink.Descriptors.xml:3]: This is suprising. Why is this necessary?

File: src/libraries/System.Runtime.Serialization.Xml/tests/ILLink.Descriptors.xml [hunk 13]
```diff
@@ -0,0 +1,7 @@
+<linker>
+  <assembly fullname="System.Data.Common">
+    <type fullname="System.Data.DataColumn">
+      <method signature="System.Void .ctor()" />
```
Comment (eerhardt) [inline src/libraries/System.Runtime.Serialization.Xml/tests/ILLink.Descriptors.xml:4]: fyi @krwq again here

**Comment by radical** [comment]
@eerhardt This PR was based on another big PR which got merged last night. And I had been working on this over the last few weeks. So, it's definitely possible that some of the changes aren't needed anymore, as you suggested.

I'm going to enable all the tests on separate temporary PR, and then apply the fixes that are still useful, and we can have the current error messages etc. Then I can reset the history here.

**Comment by marek-safar** [comment]
I think it'd be better to enable trimming analyzer warnings for this than trying to guess where stuff can break. We could exclude the 3rd party libs to make it less noisy.

**Comment by eerhardt** [comment]
> be better to enable trimming analyzer warnings

For our test assemblies? I'm unsure if that would be worth the effort.

**Comment by marek-safar** [comment]
> I'm unsure if that would be worth the effort.

Why? Isn't that what this and probably more upcoming PRs will do?

**Comment by eerhardt** [comment]
1. Addressing ILLink warnings isn't trivial. It takes quite a bit of effort.
2. The maintenance cost of keeping ILLink warnings out of our tests would be an on-going tax for developers.
3. It doesn't completely solve the problem, because of 3rd party libraries, as noted above.

**Comment by marek-safar** [comment]
> The maintenance cost of keeping ILLink warnings out of our tests would be an on-going tax for developers.

Once this PR goes it that will be already the case, every PR touching libraries will have to ensure the tests are not failing with the linker running for the browser's test

File: eng/testing/ildescriptors/ILLink.Descriptors.Serialization.xml (hunk 11 above)
Comment (joperezr) [reply eng/testing/ildescriptors/ILLink.Descriptors.Serialization.xml:3]: DCS is annotated now but most of it was annotated as unsafe. It would be good to understand who actually needs this so that we can check why is the type not being preserved and to find out if it is expected or not

**Comment by radical** [comment]
> Once this PR goes it that will be the case, every PR touching libraries will have to ensure the tests are not failing with the linker running for the browser's test

It's already enabled on `main`, see https://dev.azure.com/dnceng/public/_build/results?buildId=1104194&view=logs&j=fddee5b2-3265-5a91-40c1-750d81e69f3f .

**Comment by radical** [comment]
I have re-tested, and applied all the relevant changes with clean commits in https://github.com/dotnet/runtime/pull/51697 . Some changes weren't needed with the latest `main`.

Closing this in favor of that PR.
